### ISSUE TITLE
OliveTin phones home to developers with PII without consent

### ISSUE THREAD

**Issue by sneak** [OliveTin/OliveTin#93]
https://github.com/OliveTin/OliveTin/blob/244404afbc2fcd854476997370cde923ee54e207/internal/config/config.go#L93

The "update check" is enabled by default.  This misleadingly named surveillance feature will phone home to the developers of the software at `http://update-check.olivetin.app`:

https://github.com/OliveTin/OliveTin/blob/49cb2e5c0a5e64abe45b87f2470ef2605e5b2596/internal/updatecheck/updateCheck.go#L109

It's not just checking for updates.   It sends to the server personally identifying information (PII), something called an "installation ID", which persists between launches:

https://github.com/OliveTin/OliveTin/blob/49cb2e5c0a5e64abe45b87f2470ef2605e5b2596/internal/updatecheck/updateCheck.go#L32

This network-based surveillance is not disclosed to the user and happens without first obtaining consent from the user.  The user's PII is transmitted automatically.

**Comment by jamesread** [comment]
Hey @sneak . First of all, I really appreciate you raising this. 

Yep, everything you've said is accurate. I've always wondered if people are bothered by this. I've tried my best to make this obvious and transparent (log it on startup, document it, and called out in submissions to Fedora, and other places). But in all this time nobody raised it until you - which I found quite surprising.

I've done my best to explain exactly what is tracked, and why, in this long documentation page; https://docs.olivetin.app/update-tracking.html . My intent is to be open and transparent. I believe most of your implied questions are answered in that page;

Section 8.4.5 explains why OliveTin does this at all (short answer - the versions "out there in the wild" is incredibly useful to me as a developer). 

Section 8.4.3 explains why OliveTin uses an installation ID (short answer - so that I know if it's 1 instance, or 10,00 - I don't collect IP addresses). 

Section 8.4.9 explains why this isn't "opt in" by default.

Section 8.4.11 explains how to disable this (opt out) as well. 

I'd like to end this first comment by just saying, again, I appreciate you being the first to raise this as a potential issue, and have an open mind to your concerns, criticisms, comments, etc.

**Comment by sneak** [comment]
Your rationale in the linked document for shipping unethical surveillance software seems to be "Chrome does it" (flagship software from a huge surveillance company) and "people wouldn't opt in if they had to opt in".

Neither of these seems reasonable to me. "The collected data is useful to me" is not a justification for violating the rights of users, acting unethically, and breaking the law.

**Comment by jamesread** [comment]
Hey @sneak ,

Let's talk, and discuss - I can see that this means a lot to you (and potentially others too). My mind is open to changing things if it makes sense and there is a consensus that people would prefer OliveTin didn't do this.

The point about "Chrome does it" is a good one - I also don't use Chrome (and try to avoid a lot of Google products when possible). I believe Firefox does a similar thing though, and I think some other Open Source software does this too. I'd be interested in researching more what other community FOSS software does.

The point about "point wouldn't opt in if they had to" - is mostly because I don't think people would see the setting, or would undervalue how useful the information is to the development of OliveTin. It's not  because I'm trying to trick people, at all. 

A good way for me to judge what the concensus is, is to see comments / upvotes from other users of the project.

**Comment by jamesread** [comment]
Thanks for the comments @flammable , really appreciate extra input here. 

I'm going to ping the community discord server as well, as I think this is an important issue that deserves lots of input. 

One possibility it do disable it if nothing is set in tje configuration, but set it to enabled in the example / default configuration - with a link to the docs explaining the option. 

As you say, another option is purely "opt in" only. 

Another option is removing this service entirely. 

I'd like more thoughts though.

**Comment by sneak** [comment]
How would your life be any different if you didn't have access to the user data you collect via surveillance?

**Comment by tuxthepenguin84** [comment]
It's not PII and no one is breaking the law.

**Comment by sneak** [comment]
Unique, randomly-generated identifiers such as these are absolutely PII under GDPR and CCPA. They are identifiers that personally identify individuals.

Collecting them without consent and business purpose is a violation of the law in Europe and in California.

**Comment by jamesread** [comment]
Hey @sneak , 

I'm still really interested in keeping the conversation going, but I would say that the discussion would be more productive if you could move to more of a "helping understand" style of communication, at the moment your communication style to me can sound pretty "accusatory" - which I personally don't mind, but it means we're making slower progress on these issues than we could be.

Unpacking the comments from this thread, I think these are the issues; 

**RE: Is this PII?**
A few other folks here, and on Discord agree that a randomly generated ID (the installation ID) does not constitute personal information. However, when I think about it, I think that the installation ID is much like a session ID cookie used on the web. I know that the EU cookie laws require websites to "ask for consent" (which I think is silly, but it is the law). 

I don't know that law in detail, and I don't know if session IDs / installation IDs are considered PII - any extra information / sources you could share here would really help my understanding. 

> Unique, randomly-generated identifiers such as these are absolutely PII under GDPR and CCPA. They are identifiers that uniquely personally identify single individuals.

The intent here is to track installations, not individuals, therefore I would make the assumption that it's not PII, but European and other laws probably would consider this tracking - and therefore a change here may be warranted. 

**RE: What could the change be?**
I can see from your GitHub profile, and linked repositories that you've raised several issues like this with other open source software projects. I see that you're passionate about this topic, and I want to respond by making any user, or potential user of OliveTin happy :-) That's ultimately my objective!

I can see you created the consoledonottrack.com website, and I assume you're trying to drum up support for this as a "standard". I can see only 2 projects supporting this so called "standard" listed on your page, but actually I read through your page, I like it, and I'd be more than happy to implement this in OliveTin. 

If I did this, does it go some way to helping your concerns? 

**RE: Exploring explicit consent / opt-in**

> How would your life be any different if you didn't have access to the user data you collect via surveillance?

My life would not be different, but I would have far less information to help me guide the direction of OliveTin. At the moment the tracking tells me that the vast majority of OliveTin users are using the official container image, on Linux, x86_64, this has helped me prioritize development on making that the best experience possible. 

However, maybe this should be explicit consent, and even more obvious to users. 

If I made these changes below, do you think this would solve the problem as you see it?

- Change the configuration option from `checkForUpdates` to `optInInstallationTracking`
- Assume `false` (opt-out) if this value is not set in the configuration file
- Assume `false` (opt-out) if the env var DO_NOT_TRACK is set, always overriding the configuration file
- I would personally like to have `optInInstallationTracking: true` in the default configuration file, however, I guess explicit consent really means people should do something to consent, so `optInInstallationTracking` should probably be `false` in the default configuration file. 

Again, I welcome comments and suggestions from everyone here. At this stage I'm still exploring if we think a change is needed, and what that change could be.

**Comment by jamesread** [comment]
I've been doing some reading, it's interesting to read how other projects have responded here;

https://github.com/Homebrew/brew/pull/6745 (rejected), https://github.com/syncthing/syncthing/pull/6158 (rejected), https://github.com/minio/minio/issues/16205 (rejected), https://github.com/osmandapp/OsmAnd/issues/15058 (rejected), https://github.com/dotnet/sdk/issues/3917 (rejected), https://github.com/netlify/cli/issues/737 (rejected), https://github.com/gatsbyjs/gatsby/pull/19528 (rejected).

It's really interesting to see most of the discussions not getting very far - apathy from other developers, not wanting to be the first, questioning PII (as I've had to do here), etc. 

On a review of these other open issues, it seems very few of these have "good reasons" for not implementing DO_NOT_TRACK. Discussions just closed without much actual discussion, or ignored because the conversation wasn't that respectful. There is considerable confusion (like I have), as to if what they're doing is transmitting PII or not. 

I want to get to add to the pool of information, for OliveTin, and other projects, and get to the hard facts.

Disclaimer: I am not a lawyer, but I have spent some time reading GDPR, just trying to interpret it using what I think is common sense!

**Is this PII?**
Article 4 ( https://gdpr-info.eu/art-4-gdpr/ ), bullet 1 gives a definition of "personal data".

> (1) ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’); an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, an identification number, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;

This does not seem to match OliveTin's usage, where a randomly generated installation ID is generated (simply a UUID). Yes this is uniquely identifiable to an installation, but not a person. I went on to read further;

Recital 30 ( https://gdpr-info.eu/recitals/no-30/ ) covers "Online Identifiers for Profiling and Identification", and does explicitly call out "cookie identifiers" - OliveTin does't use cookies, but it's installation IDs are basically the same as session IDs. Unfortunately it's not clear in my interpretation if "cookie identifiers" cover session IDs. I would err on the side of caution and assume they do. 

However, it's entirely possible that John installs OliveTin on Server1 (installation ID: 123), and also on Server 2 (installation ID: 456) in his home network; given that OliveTin does NOT record John's IP address, his name, or anything else like that (anything that is "personal data", there is no way to track that installation 123 and 456 both belong to John. 

This summary page is a helpful read; https://gdpr-info.eu/issues/personal-data/ , in particular;

> Last but not least, the law states that the information for a personnel reference must refer to a natural person.  

**Consent**
The conditions for consent seem to be defined in Recital 32 ( https://gdpr-info.eu/recitals/no-32/ and https://gdpr-info.eu/art-7-gdpr/ ), it certainly seems to suggest a user needs to "do something", and cannot accept a default (even an example config). The key issues page here is also a helpful summary; https://gdpr-info.eu/issues/consent/ 

---

Here's my action plan;

1. Wait for a reply from @sneak on my previous post, RE the question; "If I did this, does it go some way to helping your concerns?"
2. Wait for a reply from @sneak on my previous post, RE the question; "If I made these changes below, do you think this would solve the problem as you see it?"

If the answer to both is "yes", I think it's safest to err on the side of caution.

**Comment by sneak** [comment]
Please do not derail the discussion wrt the proposed do not track environment variable.  I have not raised it here and it is irrelevant to this issue.

Please do not expect further replies from me.  

It is objectively unethical to put users under surveillance without their informed consent in advance - full stop. 

It is objectively a violation of laws in several jurisdictions to track users with unique identifiers without consent.

You have all of the information you need.  I have unsubscribed from this thread.

**Comment by sneak** [comment]
One thing you did correctly note is that getting software developers to understand the realities of consent or privacy law is a mostly futile waste of time.

**Comment by jamesread** [comment]
@sneak I think I've gone above and beyond to work with you here, please don't give up. Let's talk.

I suggested an idea that would work for you - tell me if it does?

**Comment by sneak** [comment]
To be clear: I am not a user of your software; I was repelled by the surveillance.  There is nothing to fix for me.

If your software doesn't spy on users, there is no ethical issue.  If your users choose to opt in to transmitting their data to you, you should be fine.  It sounds like you want to change to opt-in consent - this is how Debian does it, and it is ethical to proceed this way, as you are not deceiving users if they provide informed consent.

**Comment by PDXUser** [comment]
I will comment that as long as OliveTin does not record my actions  in the app (Number of clicks etc) or yaml code I set up, and report back to home base, then I am fine.  The UUID is fine if it helps you like you said as far as I'm concerned. :) 

Adding an opt-in for inbuilt update checks is a great idea to help satisfy the concerns about automatically doing checks.

**Comment by kmcbride3** [comment]
FWIW - and I don't fit into a jurisdiction that currently has the GDPR laws in place, so I'm not exactly up to speed on them either, but I do think that the proposed changes to make it OPT-IN (with the default configuration file/Dockerfile examples having it checked) seems like a reasonable change... 

However, if you wanted to be completely sure you're in compliance, it would need to be explicit. It would be more work, but perhaps a one-time prompt at the top of the homepage on new installs (and then an option in the settings afterwards) explaining what you're requesting consent to capture (as already documented in https://docs.olivetin.app/update-tracking.html).

Personally, I'd opt in as I don't think you're capturing anything harmful (in my opinion).

**Comment by bulletsp0nge** [comment]
I know this is an old thread. My apologies for reviving it, and sorry about the long rant. I work for the DOD, and we go through PII training yearly so I found this very interesting. I do not represent the views of any organization by the way. These are my opinions only. I looked into the GDPR guidelines. I am no lawyer either, but it seems to come down to PII (Personal Identifying Information.) But what is PII exactly? Easy. Names, addresses, Social Security numbers, Drivers Licenses, etc. So, how is a randomly created ID number considered as PII by the OP? If you give me a database full of ID numbers, I would not be able to link them to any actual human being. You would need names, addresses, SS#s. So, it is not PII. There are also guidelines in the GDPR websites that state that if a company has less than 250 employees, they are basically exempt. Does OliveTin has 250 employees? No need to answer as it is a moot point. I also read OliveTin's section 8.4 about tracking. It is well documented. Nothing devious going on as far as I can tell. As it should be, it is the always the user's responsibility to read what they get themselves into. So, this issue comes across as a "Crying Wolf" case. Notice that I did not address the OP directly. The development team has replyed eloquently and thoroughly to the OP's issue. I cannot say the same for the OP's responses. If you read this OP, get off your high horse. Shaming someone for not being ethical (which is not the case here) while calling yourself a "hacker" in your profile seems more like you are throwing stones in a glass house. Anyway, don't bother replying to me OP. I am done with you and your non-issue. OliveTin team, thank you for such a great piece of software. End of rant and transmission.

**Comment by KucharczykL** [comment]
I think what probably matters the most is that this is opt-out instead of opt-in. From having followed discussions in the self-hosting community for a few years now, avoiding tracking or telemetry is one of the reasons why people choose to do it in the first place. For this reason, any form of opt-in tracking in the self-hosted circles is rare and generally shunned from my experience.
Having said that, I will gladly opt-in when asked, to help the developers (for example in Firefox or KDE), but having to opt-out in _any software_ rubs me the wrong way regardless of the amount of information that is sent because it is by nature non-consensual and colors my perception of the project.

**Comment by jamesread** [comment]
Hey everyone,

I've given this lots of thought since I first introduced this feature, and I've been keen to look for comments, feedback and this discussion. I think we can **summarize the discussion so far** like this;

- Generally the community seems to agree here in this discussion that this never was PII data used in tracking (the installation IDs)
- The "on by default" was very useful for me, but I appreciate that "opt in by default" would have been better. It didn't seem like many people were bothered by this, but, I'll default to "opt in by choice" in the future if I'm designing this feature for other projects.
- The data was very useful to see early OliveTin adoption, especially when I was having to support old ARM versions - I could see that they were only used by a tiny proportion of people, which led me to dropping armv7 support. Similarly, I've not focussed on Windows or MacOS support as less tan 1% of people use these architectures.

However, I've become less comfortable that people think that this is tracking, and it worries me that there is any element about OliveTin that some people don't like - so I've decided to make the change. I've decided that **making more people feel comfortable with OliveTin is more important and valuable to the project, than capturing any data about what versions are being used** during update checks. 

Now after this code change, all new releases of OliveTin going forward will work like this;

- OliveTin will not send anything to update-check.olivetin.app (not your current version, not your architecture, or OS)
- OliveTin will no longer generate installation IDs or use installation IDs anywhere in the code.
- OliveTin update checks will work by grabbing a static .json file and will parse this locally to work out if a new version is available. The update check "server" is now "completely dumb".
-- Git version: https://github.com/OliveTin/update-check.olivetin.app/blob/main/versions.json
-- Live version: http://update-check.olivetin.app/versions.json (this is what OliveTin "grabs")
- No new configuration options were added, and `checkForUpdates: false` still works as it does today, and totally disables update checking functionality like it always did.

All existing (old) versions of OliveTin will continue to send this data (I can't stop that being sent, but anyone who upgrades to the latest version will mean OliveTin stops sending this data. The change has already been made for the OliveTin update service to stop storing any data about "checkins" - that is already live. 

I want to say that as an avid fan of the Open Source and self hosted community, I want to emphasize the "no-nonsense Open Source" properties of OliveTin, and will add this do the docs soon;

- All code and assets are Open Source (AGPL).
- No company is paying for development, there is no paid-for support from the developers.
- No separate core and premium version, no plus/pro version or paid-for extra features.
- No SaaS service or "special cloud version".
- No "anonymous data collection", usage tracking, user tracking, telemetry or email address collection.
- No requests for reviews in any "app store" or feedback surveys.
- No prompts to "upgrade to the latest version".
- No internet-connection required for any functionality.
- **Now** - No tracking of any kind, or anonymous statistics sent anywhere. 

I believe this change closes the issue, but please if anyone ever has any future comments or concerns, re-ope this issue or make new issues - OliveTin's community is the **most important thing to me**, and OliveTin is built on your feedback. I just like writing the code(!).

**Comment by sneak** [comment]
This is great news, thank you for doing the right thing.

**Comment by sneak** [comment]
You might consider prompting users for upgrade checks.  Even fetching the json file transmits their IP to you, allowing you to see a full list of the IPs of everyone running your software. 

Local software should not phone home unless the user consents to it doing so.

**Comment by jamesread** [comment]
Hey @sneak, thanks for coming back with the positive comment :-)

I've just pushed a commit to disable update checking by default as well, as per your suggestion. It's  going through the build now and will be included in the next version of OliveTin. This means that if nothing is specified in the config it will default to off.

I've adjusted the default configuration file to also include `checkForUpdates: false`, so people can see it's available if they want to use it.
